clang/include/clang/Basic/DiagnosticSemaKinds.td

```diff
@@ -5663,6 +5663,8 @@ def err_explicit_instantiation_internal_linkage : Error<
 def err_explicit_instantiation_not_known : Error<
   "explicit instantiation of %0 does not refer to a function template, "
   "variable template, member function, member class, or static data member">;
+def err_explicit_instantiation_no_candidate : Error<
+    "no candidate for explicit instantiation of %0">;
```

Comment:
Can't you use the above `err_explicit_instantiation_not_known` diagnostic for this? It's is already used in a similar scenario when the member functions are templates: https://godbolt.org/z/bToW3o46P (Or is this used for member function templates as well now?)

Reply:
I think we should the new diagnostic in more places. We did find a matching entity, saying it doesn't exist is not super useful

Reply:
This would be a fairly significant change, are you opposed to do that as a separate PR?
`err_explicit_instantiation_not_known` basically never explain what the issue is (no notes)

Reply:
There's notes in the example above, but yeah no reason to change everything in this PR when it seems unrelated.  
This message should say "no *viable* candidate" (or "no matching function").

Also you should be able to restructure stuff so that template/non template member functions can use the same overload set, so would naturally use the same diagnostic. For example, this:

```c++
template<typename>
struct A {
  template<typename = void>
  void f() requires(false);
  void f() requires(false);
};
template void A<int>::f();
```

Should note both candidates, when I think your change means it only notes the non-template candidate (and removing the non-template function means it notes the template one, which seems strange)

Reply:
Agreed, but that's a bigger refactor